### PR TITLE
[feature] AntiSpam improvements 

### PR DESCRIPTION
**Pull request by Nathan-Webb**
**Describe the pull**
Added a new chat module that censors or removes slurs and swears. It also attempts to remove any custom suffixes and unicode-abusing fancychat.

**Describe how this pull is helpful**
Allows for a bit more fine-tuning in terms of chat filtration. Removing most suffixes seems like a good quality of life improvement.

**Additional context**
Should there be a check to ensure that the "Experimental" setting is enabled before running the fancychat/suffix sanitizer?

### CODE REVIEW

**Comment by dominikaaaa** [comment]
I think this should be combined with AntiSpam, and most of the AntiSpam settings can be removed since they were added before ChatFilter was a module. I'll take a look at this later, thanks

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatSanitizer.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            replacement = s;
+        }
+        public String getReplacement() {
+            return replacement;
+        }
+
+    }
+
+    //fairly simple - anarchy servers shouldn't have a chat filter so there isn't an attempt to check for bypasses
+    private static final String[] slurMatchers = new String[]{"nigg(a|er)", "chink", "trann(y|ie)", "kike", "fag(got)?", "retard(ed)?"};
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatSanitizer.java:110]: I'd rather obf these regexes so they're not in the source. I'll do it myself later.

Comment (Nathan-Webb) [reply src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatSanitizer.java:110]: why obfuscate the regexes? It's just meant to be a quick and dirty catch for most cases. I highly doubt someone is going to try and specifically bypass this. If thats what you want with this module then I believe there are a few java libraries that do a better job at slur detection than a handful of simple regexes. 😛

Comment (dominikaaaa) [reply src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatSanitizer.java:110]: It doesn't look good to have it in the source as a searchable thing, even if the intention is to remove it from the chat.

Comment (Nathan-Webb) [reply src/main/java/me/zeroeightsix/kami/module/modules/chat/ChatSanitizer.java:110]: oohhh I misinterpreted your message. I see your point now. Good idea!

**Comment by dominikaaaa** [comment]
I integrated it into antispam as earlier, but I just realized that you made this pull before https://github.com/kami-blue/client/commit/2e221c450b78f3dbb117d2efa01cff01f1f2dda2 was merged. You'd have to reimplement that.

Another weird thing, messages that aren't detected for anything period are cancelled somewhere?

**Comment by Nathan-Webb** [comment]
>You'd have to reimplement that.
Oof, missed that one. Will do!

> Another weird thing, messages that aren't detected for anything period are cancelled somewhere?
I'll look into that. 90% sure I just left debug code somewhere

**Comment by Nathan-Webb** [comment]
Should I preserve the options of censoring the text instead of outright cancelling the message when a slur/swear word is detected? It would just involve moving `put(slurs, FilterPatterns.SLURS)` and  `put(swears, FilterPatterns.SWEARS)` in their own list, and checking separately.

**Comment by dominikaaaa** [comment]
Why not have it in one list?

**Comment by Nathan-Webb** [comment]
because in one list I can't check if the thing detected was a spammy phrase or a slur without making some modifications to the detectSpam function which I feel would make the code a bit more complicated. splitting it up into a `detectSpam` and `sanitizeNastyWords` would allow me to implement censoring options. 

Currently the code cancels the message if it found a nasty word and a spammy phrase. I feel that keeping an option to censor instead of just hiding the message would be beneficial.

**Comment by dominikaaaa** [comment]
Why not just modify this to censor the slurs? That's why I made the `String?` changes. 

It looks like part of my changes didn't get pushed :/

So here we should be returning `String?` instead and null for false:

Screenshot_20200826_120422

And then here we would be checking if is spam is null, and if it isn't null, replace the result of isSpam with the censoring filter, if you know what I mean?

Screenshot_20200826_120243

**Comment by Nathan-Webb** [comment]
ohhh i totally missed those. I blindly wrote over those changes because I thought it was part of the conflict. I'll set that up soon, its too early for me. 😛

**Comment by wnuke** [comment]
For fancy chat/small caps you can actually just subtract a number from each "fancy" character to convert it to a "normal" character. For example subtract 120377 from a mathematical monospace capital a (𝙰) to get a latin (normal) capital letter a (A).

**Comment by Nathan-Webb** [comment]
Oooh I did not realize that. I'll test that out and add that alongside my changes.

**Comment by Nathan-Webb** [comment]
Apologies, but I'm wracking my brain trying to figure out how to implement the duplicate detection alongside the `Replace` Mode.

Do I just cancel the message regardless of the selected mode if a duplicate message is received? Or should I replace the duplicate message with something else. I don't really know how do combine the two.

My current duplicate checking. 

```
if (duplicates.value) {
            if (messageHistory == null) messageHistory = ConcurrentHashMap()
            var isDuplicate = false

            if (messageHistory!!.containsKey(message) && (System.currentTimeMillis() - messageHistory!![message]!!) / 1000 < duplicatesTimeout.value) isDuplicate = true
            messageHistory!![message] = System.currentTimeMillis()

            if (isDuplicate) {
                sendResult("Duplicate", message)
                return ".*" // should this be returned so that the entire message is blanked out, and have a check somewhere else which cancels the message if the resulting text is blank? 
            }
        }```

**Comment by dominikaaaa** [comment]
If it's a duplicate, cancel it entirely.

**Comment by Nathan-Webb** [comment]
alrighty, sounds good

**Comment by Nathan-Webb** [comment]
Would I be able to add a check here? It seems a bit finicky if I add a way to cancel the event in the `detectSpam` function.

Something like this
```
val pattern = isSpam(event.message.unformattedText)
        if (pattern != null) { // null means no pattern found
            if (mode.value == Mode.HIDE) {
                event.isCanceled = true
            } else if (mode.value == Mode.REPLACE) {
                event.message = TextComponentString(sanitize(event.message.formattedText, pattern, (replaceMode.value as ReplaceMode).redaction))
            }
        }
        //check for duplicate here ```

**Comment by dominikaaaa** [comment]
yea, go for it

**Comment by Nathan-Webb** [comment]
To resolve the merge conflict should I just squash merge with master then push to the branch? Or are you just going to manually edit the imports when the time comes. I feel like its at a good point where it can be pushed to main, but I have a question. Should `FilterPatterns.GREETER` include vanilla join/leave? IMO join/leaves (at least the vanilla / essential server message) should be separate from the automated greeters that clients use.

**Comment by dominikaaaa** [comment]
I'll fix conflicts myself. 

> Should FilterPatterns.GREETER include vanilla join/leave? IMO join/leaves (at least the vanilla / essential server message) should be separate from the automated greeters that clients use.

Yes, most people don't want to see them if they enable the setting

**Comment by Nathan-Webb** [comment]
Sounds good!

> Another weird thing, messages that aren't detected for anything period are cancelled somewhere?

I couldn't seem to repeat this bug during my testing. Everything else seems to be in order and ready for your review.

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt [hunk 2]
```diff
@@ -79,8 +89,27 @@ class AntiSpam : Module() {
...
+            } else if (mode.value == Mode.REPLACE) {
+                event.message = TextComponentString(sanitize(event.message.formattedText, pattern, (replaceMode.value as ReplaceMode).redaction))
+            }
+        }
+
+        if (duplicates.value) {
+            if (checkDupes(event.message.unformattedText)) {
+                event.isCanceled = true
+            }
+        }
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt:106]: I'd favor duplicate checking over censoring, can you move this before the pattern check?

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt [hunk 3]
```diff
@@ -32,37 +33,46 @@ import java.util.stream.Collectors
...
+    private val swears = register(Settings.booleanBuilder("Swears").withValue(false).withVisibility { p.value == Page.TYPE }.build())
+    private val automated = register(Settings.booleanBuilder("Automated").withValue(true).withVisibility { p.value == Page.TYPE }.build())
+    private val ips = register(Settings.booleanBuilder("ServerIps").withValue(true).withVisibility { p.value == Page.TYPE }.build())
+    private val specialCharEnding = register(Settings.booleanBuilder("SpecialEnding").withValue(true).withVisibility { p.value == Page.TYPE }.build())
+    private val specialCharBegin = register(Settings.booleanBuilder("SpecialBegin").withValue(true).withVisibility { p.value == Page.TYPE }.build())
+    private val greenText = register(Settings.booleanBuilder("GreenText").withValue(false).withVisibility { p.value == Page.TYPE }.build())
+    private val experimental = register(Settings.booleanBuilder("Experimental").withValue(false).withVisibility { p.value == Page.TYPE }.build())
+    private val fancyChat = register(Settings.booleanBuilder("FancyChat").withValue(false).withVisibility { experimental.value && p.value == Page.TYPE }.build())
```
Comment (dominikaaaa) [inline src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt:51]: Just get rid of the experimental setting entirely

**Review by dominikaaaa — CHANGES_REQUESTED** [review]

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt (hunk 2 above)
Comment (Nathan-Webb) [reply src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt:106]: sounds good!

File: src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt (hunk 3 above)
Comment (Nathan-Webb) [reply src/main/java/me/zeroeightsix/kami/module/modules/chat/AntiSpam.kt:51]: will do!

**Review by dominikaaaa — APPROVED** [review]

**Review by dominikaaaa — APPROVED** [review]
Everything looks good to me

**Comment by dominikaaaa** [comment]
I need to fix conflicts

**Comment by dominikaaaa** [comment]
```
error: failed to push some refs to 'github.com:Nathan-Webb/client.git'
To github.com:Nathan-Webb/client.git
!	refs/heads/fix/AntiSpam:refs/heads/fix/AntiSpam	[remote rejected] (permission denied)
Done
```
Actually, you disabled the option that allows me to push to your pull. You have to fix it yourself @Nathan-Webb

**Comment by Nathan-Webb** [comment]
wtfrick? I thought I kept that option enabled. Its just the little checkbox right?

**Comment by dominikaaaa** [comment]
Yes

**Comment by dominikaaaa** [comment]
If you enable it I can push the fixed version

**Comment by Nathan-Webb** [comment]
image
Am I looking at the wrong thing?

**Comment by dominikaaaa** [comment]
That should be it. It's still not letting me push, you'll either have to add me to your repo or fix the conflicts yourself I guess

**Comment by Nathan-Webb** [comment]
hmm, is the account in your git/intellij instance the same as this one? I'll add you to the repo in a moment

**Comment by Nathan-Webb** [comment]
there we go. invite pending :thonkkill:

**Comment by dominikaaaa** [comment]
Yes.

**Comment by dominikaaaa** [comment]
That worked I guess. Weird

**Comment by dominikaaaa** [comment]
I'll finish this tomrrow

**Review by l1ving — APPROVED** [review]
Looks good, sorry for the delay. Thank you!

**Pull request merged**